### PR TITLE
Fix checkpoint loading with `load_main_params_from_ckpt=True` for grouped weight

### PR DESCRIPTION
**Pull request by ksivaman**
# What does this PR do ?

Support loading a multi-weight `GroupedLinear` checkpoint into a single weight module and vice-versa. This affects the path where the optimizer params directly load the checkpoint. The normal (model) loading path is already supported.

### Pre-checks

- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

**Comment by kunlunl** [comment]
The placement of the `_normalize_state_dict_for_grouped_params` call looks good to me. And if I understand correctly, the initial check in the function — `if not (hasattr(module, 'num_gemms') and hasattr(module, '_split_grouped_checkpoint_tensor')` — ensures that cases without grouped weights remain unaffected. Is that correct?

So overall above part looks good to me. But since I’m not familiar with the grouped weight checkpoint format, is there anyone else can review the internal logic of the `_normalize_state_dict_for_grouped_params` function @yaox12 ?

**Review by yaox12 — APPROVED** [review]
